R/calculate_study_day.R

```diff
@@ -0,0 +1,132 @@
+#' example, AESTDY for AE domain and CMSTDY for CM domain.
+#' @param merge_key Character to represents the merging key between `sdtm_in` and
+#' `dm_domain`.
+#'
+#' @return Data frame that takes all columns from `sdtm_in` and a new variable
+#' to represent the calculated study day.
+#'
+#' @export
+#'
+#' @examples
+#' ae <- data.frame(
+#'   USUBJID = c("study123-123", "study123-124", "study123-125"),
+#'   AESTDTC = c("2012-01-01", "2012-04-14", "2012-04-14")
+#' )
+#' dm <- data.frame(
+#'   USUBJID = c("study123-123", "study123-124", "study123-125"),
+#'   RFSTDTC = c("2012-02-01", "2012-04-14", NA)
+#' )
+#' ae$AESTDTC <- as.Date(ae$AESTDTC)
+#' dm$RFSTDTC <- as.Date(dm$RFSTDTC)
+#' calculate_study_day(ae, dm, "AESTDTC", "RFSTDTC", "AESTDY")
+#'
+calculate_study_day <- function(sdtm_in,
+                                dm_domain,
+                                tgdt,
+                                refdt,
+                                study_day_var,
+                                merge_key = "USUBJID") {
+  assertthat::assert_that(is.data.frame(sdtm_in))
+  assertthat::assert_that(is.data.frame(dm_domain))
+  assertthat::assert_that(
+    utils::hasName(dm_domain, refdt),
+    msg = "dm_domain needs to have the variable of refdt."
+  )
+  assertthat::assert_that(
+    utils::hasName(sdtm_in, tgdt),
+    msg = "sdtm_in needs to have the variable of tgdt."
+  )
+  assertthat::assert_that(
+    utils::hasName(dm_domain, merge_key),
+    msg = "dm_domain needs to have the variable of merge_key."
+  )
+  assertthat::assert_that(
+    utils::hasName(sdtm_in, merge_key),
+    msg = "sdtm_in needs to have the variable of merge_key."
+  )
+  assertthat::assert_that(is.character(study_day_var))
+  # check tgdt and study_day_var matching, for example, CMSTDTC matches CMSTDY
+  if (gsub("DTC", "", tgdt, fixed = TRUE) != gsub("DY", "", study_day_var, fixed = TRUE)) {
+    warning(
+      "Target date and the returned study day doesn't match. ",
+      "Expecting matching date and study day, for example, CMENDTC and CMENDY"
+    )
+  }
+
+  original_variables <- names(sdtm_in)
+
+  if (!identical(sdtm_in, dm_domain)) {
+    dm_domain <- unique(dm_domain[c(merge_key, refdt)])
+
+    check_refdt_uniqueness <- dm_domain |>
+      dplyr::group_by(dplyr::pick({{ merge_key }})) |>
+      dplyr::filter(dplyr::n() > 1L)
+    if (nrow(check_refdt_uniqueness) > 0L) {
+      warning(
+        "Reference date is not unique for each patient! ",
+        "Patient without unique reference date will be ingored. ",
+        "NA will be returned for such records."
+      )
+      dm_domain <- dm_domain[
+        !dm_domain[[merge_key]] %in% check_refdt_uniqueness[[merge_key]],
+      ]
+    }
+
+    sdtm_in <- sdtm_in |>
+      dplyr::left_join(
+        dm_domain,
+        by = structure(names = merge_key, .Data = merge_key)
+      )
+  }
+
+  # refdt/tgdt should be in ISO format, otherwise throw warning
```

Comment:
More than being in ISO format, they must be in a specific ISO format, i.e. `"%Y-%m-%d"`.

I would split the code that follows (i.e. the `tryCatch()` and `as.Date()` part) in two separate portions.

- Firstly, before the `if (!identical(sdtm_in, dm_domain))`, I would perform an assertion to check that indeed `sdtm_in[[refdt]]` and `sdtm_in[[tgdt]]` are formatted as  `"%Y-%m-%d"`, or warn. I would write a dedicated function for this task that takes each vector, e.g something like `assert_ymd(x, .warn = TRUE)`.

- Secondly, I would use a function for the actual conversion from ISO format `"%Y-%m-%d"` to a Date object, e.g.:
  ```R
  ymd_to_date <- function(x) {

    admiraldev::assert_character_vector(x)
    dates <- as.Date(x, "%Y-%m-%d", optional = TRUE)

    dates
  }
  ```

Reply:
Hi @ramiromagno this is supposed to be a very easy function, and I would prefer not to have have so many separate small functions. I will apply the assertion logic in the body of the main function instead. Thanks!

Reply:
Hi @ramiromagno so I thought about it and I don't think I should add the admiraldev::assert_character_vector(x) here. What if the input reference date and target date are already dates?